Added new Fluid hooks; #1643

litzinger opened the pull request:
resolves #1643

  - Added fluid_field_after_update_field hook
  - Added fluid_field_after_add_field hook
  - Added additional parameters to fluid_field_get_field_data and fluid_field_add_field hooks
  - Move fluid_field_get_all_data hook, it was in the wrong place
  - Only attempt to save Fluid data if savable data exists

These are changes and additions made to hopefully allow Publisher to support Fluid field.

The addition of the conditionals checking if $values is empty or not lets the first hook return no data. In Publisher if saving a non-default language, we don't want the native Fluid field to be updated with non-default data.

The change in the foreach loop when displaying fields is hard to explain at this point because it's been so long since I originally tested it, but it seems like a logical change, even outside the context of these hook changes. I tested it with and without Publisher installed and it still saves and displays Fluid field data just fine.

I'm pretty certain that the fluid_field_get_all_data hook was just in the wrong file. There are two getFieldData methods in different files and I think it was some confusion when the hook was originally added.

With these changes I'm able to successfully save basic text fields in Fluid with drafts and translations. Default Fluid functionality remains unchanged. Grid and Relationship fields inside of Fluid are a different beast that I need to figure out, but I'm pretty certain those need to be managed outside the context of any hooks added to the Fluid field.

litzinger committed:
Add Fluid hooks and new parameters to existing hooks

litzinger committed:
Move the fluid_field_get_all_data hook

It was incorrectly in the getFieldData method that fetches data for a _single_ field inside a Fluid field, not the entire Fluid field collection.

litzinger committed:
Return functionality

litzinger committed:
Reenable the empty array check

intoeetive commented on system/ee/ExpressionEngine/Addons/fluid_field/ft.fluid_field.php:
```diff
@@ -410,10 +410,21 @@ private function updateField($fluid_field, $order, $group, array $values)
-        $query->set($values);
-        $query->where('id', $fluid_field->field_data_id);
-        $query->update($fluid_field->ChannelField->getTableName());
+        if (!empty($values)) {
```
Worried about this check, will it not block removing all fields from Fluid to save it as empty?

Also similar one below

litzinger replied:
No b/c that's handled by removeField(). I just did a test without Publisher installed just to verify, I was able to remove all fields from a Fluid field, saved, and it reloaded with an empty Fluid field.

intoeetive commented on system/ee/ExpressionEngine/Addons/fluid_field/ft.fluid_field.php:
```diff
@@ -593,17 +620,22 @@ public function display_field($data)
                     } else {
-                        $field = $field_data[0]->getField();
+                        foreach ($field_data as $field_datum) {
+                           $field = $field_datum->getField();
```
Please see PSR-12 notices on files tab in the PR, looks like you're missing some spaces

litzinger replied:
Whoops.

intoeetive reviewed (changes requested):
@litzinger thank you for submitting this!

Please see a couple of comments I posted on the code. 

Am I getting it right that including this PR is making #3927 unnecessary and merging it will close #1643 ?

Also asking @bryannielsen to review

litzinger committed:
Fix spacing

litzinger commented:
> Am I getting it right that including this PR is making #3927 unnecessary and merging it will close #1643 ?

That is correct.

litzinger commented:
Hold off on merging this... I think I might need to add a hook to the very end of `Fluid_field_ft->save()` or `post_save()`

bryannielsen commented:
@litzinger I'm going to try and get some time into reviewing this soon since it's related to work I've been doing on #4030.  Are there more changes you're wanting to make before I take a look?

litzinger commented:
> @litzinger I'm going to try and get some time into reviewing this soon since it's related to work I've been doing on #4030. Are there more changes you're wanting to make before I take a look?

At this point no, I got it working with Relationship fields without further changes. 🤞 the Grid implementation goes smoothly.

litzinger committed:
Exploring a different option

litzinger commented:
I got Grid and Relationship field saving correctly, then moved onto the sync'ing feature that Publisher has (saving as published will update the draft content to match the published data), and when saving a draft as published, all heck breaks loose because it is not properly removing existing published data to replace it with the draft data. The very specific hooks don't give me a broader picture of what is happening when a save is occurring, so I'm exploring a new option, which is a before and after save hook. The Grid field does something similar by building a collection of added, removed, and updated rows. Maybe that'll work with Fluid as well. It'll take me a few days to get a clearer picture if this route will work instead of the previously suggested hooks.

https://github.com/ExpressionEngine/ExpressionEngine/pull/4125/commits/5413d8d21015c539318a179339fae3c50fba6c05

litzinger committed:
Change key names and alphabetically order them

bryannielsen committed:
Revert changes to rendering in display_field()
